Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Utilities.ps1

```diff
@@ -232,24 +232,172 @@ function Get-ServiceFabricApplicationAction
+            $appType = Wait-ServiceFabricApplicationTypeStatusChange -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion
+        }
+
+        # if app type got unprovisioned, retry register
+        if(!$appType)
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeUnprovisioned)
+            return $true
+        }
+
+        # if app type is provisioned, bail out
+        if ($appType.Status -eq [System.Fabric.Query.ApplicationTypeStatus]::Available)
+        {
+            Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeProvisioned)
+            return $false
+        }
+
+        # if provisioning failed, throw and don't retry
+        throw (Get-VstsLocString -Key SFSDK_RegisterAppTypeFailedWithStatus -ArgumentList @($appType.Status, $appType.StatusDetails))
+    }
+
+    Invoke-CommandWithRetries -Action $registerAction `
+        -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingRegisterApplicationType) `
+        -OnException $onException
+
+    Write-Host (Get-VstsLocString -Key SFSDK_ApplicationTypeProvisioned)
+}
+
 function Get-ServiceFabricApplicationTypeAction
 {
     Param (
         [string]
-        $ApplicationTypeName
+        $ApplicationTypeName,
+
+        [string]
+        $ApplicationTypeVersion
     )
 
     $global:operationId = $SF_Operations.GetApplicationType
-    return Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName
+    $getApplicationTypeParams = @{
+        'ApplicationTypeName' = $ApplicationTypeName
+    }
+
+    if ($ApplicationTypeVersion)
+    {
+        $getApplicationTypeParams['ApplicationTypeVersion'] = $ApplicationTypeVersion
+    }
+
+    $getAppTypeAction = { Get-ServiceFabricApplicationType @getApplicationTypeParams }
+    return Invoke-CommandWithRetries -Action $getAppTypeAction `
+        -RetryMessage (Get-VstsLocString -Key SFSDK_RetryingGetApplicationType)
 }
 
-function Get-ServiceFabricApplicationUpgradeAction
+function Wait-ServiceFabricApplicationTypeStatusChange
 {
     Param (
         [string]
-        $ApplicationName
+        $ApplicationTypeName,
+
+        [string]
+        $ApplicationTypeVersion
     )
 
-    $global:operationId = $SF_Operations.GetApplicationUpgradeStatus
-    return Get-ServiceFabricApplicationUpgrade -ApplicationName $ApplicationName
+    $global:operationId = $SF_Operations.GetApplicationType
+    $getAppTypeAction = { Get-ServiceFabricApplicationType -ApplicationTypeName $ApplicationTypeName -ApplicationTypeVersion $ApplicationTypeVersion }
+    $appTypeStatusValidator = {
+        param($appType)
+        # if app type does not exist (i.e it got unprovisioned) or if its status has changed to a terminal one, stop wait
+        if ((!$appType) -or (($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Provisioning) -and ($appType.Status -ne [System.Fabric.Query.ApplicationTypeStatus]::Unprovisioning)))
```

Comment:
this statement doesn't look correct

Reply:
this is correct. if app type is in terminal state, then retry should not be attempted. This is ActionSuccessValidator delegate and hence $true means no further retry.